### PR TITLE
firewalld: Add support for default_zone configuration

### PR DESCRIPTION
**Pull request by spacekpe**

### CODE REVIEW

**Comment by ghost** [comment]
you're gonna need to add a "version_added:" to the new argument if you want this to be accepted as quick as possible. currently that would be 1.9

**Comment by spacekpe** [comment]
@jrobeson , Done.

**Comment by gregdek** [comment]
@spacekpe Thanks for submitting this pull request. Unfortunately, it appears that the underlying code has changed since your submission, and this PR now needs to be rebased. 

If you believe that this PR still applies, please rebase and we will put it into our community review process as soon as possible.

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action, and it hasn't been updated in the last month. 

If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we'll keep it open pending your changes.

If we receive no indication from you within a week, we will close this pull request.

**Comment by gregdek** [comment]
Thanks @spacekpe -- now needs a bump to the latest version tag, which is now 2.0.

When you're done, please add a comment with the text "ready_for_review" and we will continue with the review process.  Thanks!

**Comment by gregdek** [comment]
Also: @maxamillion, maybe you can take a look at this PR while we're waiting for @spacekpe to change the version tag.

Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by maxamillion** [comment]
:+1:

**Comment by robynbergeron** [comment]
@spacekpe -- looks like all we need is the change @gregdek mentioned (version tag bump to 2.0), and this will be ready to merge! Sorry this has been so slow in moving, but the good news is that @gregdek and I are working to improve the process and catch up on backlog... and the better news is that this is SUPER CLOSE to going in now. 

Thanks, @maxamillion, for the review. :)

**Comment by spacekpe** [comment]
@robynbergeron, Version bump done.

**Comment by robynbergeron** [comment]
@spacekpe -- thanks for that. Since @maxamillion gave his thumbs up and we have the version tag fixed -- moving this to shipit for inclusion! 

Thanks to @spacekpe in general for your patience. Feels SOOOOOOO GOOD to finally process things submitted so long ago :)

cc @ansible/core team

**Comment by gregdek** [comment]
So wait... now this needs a rebase?

I'm confused, because the last commit to this file, according to Github, was 24 days ago:

https://github.com/ansible/ansible-modules-extras/commit/39eb3807f3497fffeea19183691b54489a811239

I'm wondering if the commit to add the version bump landed in a conflicting branch... or something?

@spacekpe can you rebase so this applies cleanly? I promise we'll get it committed asap when we see that there are no merge conflicts.

**Comment by gregdek** [comment]
Also @spacekpe please indicate that you've rebased by putting "ready_for_review" in the comments. 

Sorry for the delay -- so close to the finish line here. :/

**Comment by pspacek** [comment]
ready_for_review

@gregdek , Here it is! I finally had time to apply the patches.

At the same time I found out that default zone setting is always permanent and immediate change, so now the `default_zone` option must be used with `permanent = True` and `immediate = True` so the user is aware of the fact.

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by pspacek** [comment]
@gregdek I already did the changes and as far as I can see, there is was no reply to latest version of this PR. Is there some problem with it?

**Comment by gregdek** [comment]
Thanks @spacekpe. @maxamillion please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by maxamillion** [comment]
LGTM :+1:

**Comment by gregdek** [comment]
Thanks again to @spacekpe for this PR, and thanks @maxamillion for reviewing. Marking for inclusion.

File: system/firewalld.py [hunk 1]
```diff
@@ -66,14 +66,19 @@
   state:
     description:
       - "Should this port accept(enabled) or reject(disabled) connections."
-    required: true
+    required: false
```
Comment (bcoca) [inline system/firewalld.py:76]: why this change? also no default stated or added

File: system/firewalld.py [hunk 2]
```diff
@@ -277,6 +283,11 @@ def main():
     else:
         zone = fw.getDefaultZone()
 
+    if module.params['default_zone']:
```
Comment (bcoca) [inline system/firewalld.py]: this is redundant, just use `default_zone = module.params['default_zone']`

**Comment by gregdek** [comment]
Thanks @spacekpe for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

File: system/firewalld.py (hunk 1 above)
Comment (pspacek) [inline system/firewalld.py:76]: Well, parameter state does not make sense when changing default zone. That was the reason why I made it optional. Does it answer your question? How do you propose to document it?

**Comment by gregdek** [comment]
Thanks @spacekpe for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@spacekpe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by pspacek** [comment]
ready_for_review

File: system/firewalld.py [hunk 3]
```diff
@@ -401,9 +408,20 @@ def main():
...
+        module.fail_json(msg='can only operate on port, service, rich_rule, '
+                         'interface or default_zone at once')
+
+    if (default_zone is None) == (desired_state is None):
+        module.fail_json(msg='mandatory options are either (default_zone, '
+                         'permanent, immediate) or (state, permanent)')
+
+    if default_zone and (not permanent or not immediate):
+        module.fail_json(msg='default_zone can be used only with options '
+                         'permanent and immediate set to True')
```
Comment (maxamillion) [inline system/firewalld.py:511]: I'd like to see a check here that makes sure that `desired_state` has a non-`None` value for all options other than `default_zone` so that for all other options that aren't using `desired_state` the functionality is the same (i.e. - it would fail in the event that `state` parameter is omitted)

Maybe something along the lines of:

``` python
if (default_zone is None) and (desired_state is None):
    module.fail_json(msg='"state" is a mandatory option when not performing an operating on "default_zone"')
```

Comment (pspacek) [reply system/firewalld.py:511]: I believe that this is already done on line 418 (one if above the one you commented on). Am I wrong?

Comment (maxamillion) [reply system/firewalld.py:511]: Oversight on my part. Apologies.

**Comment by maxamillion** [comment]
:+1: shipit

**Comment by gregdek** [comment]
Thanks again to @spacekpe for this PR, and thanks @maxamillion for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @spacekpe for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by maxamillion** [comment]
Shouldn't the bot have merged this 10 days ago?

CC: @bcoca @jimi-c @jctanner

**Comment by pspacek** [comment]
Rebased. Please merge it.

**Comment by maxamillion** [comment]
shipit

**Comment by gregdek** [comment]
Sorry about this one @spacekpe -- ugh. Shippable fails because we've set the "merge" version to 2.3 here, even though it first reached "shipit" status when we were still clearly in the 2.2 window. So looks like we missed this one, again.

I'm putting into "shipit" manually and marking "bot_broken" so Shippable doesn't tag it incorrectly again. My advice is that whoever merges just manually changes the version_added field as needed. We've been fighting to get this one in for almost 2 years at this point.
